In [1]:
# Part 1: Starting Simple - Create Initial Prompts
# Objective: Create basic zero-shot prompts for three different task types and establish a baseline.

# Step 1: Setup Your Environment
# What to do:

# Create a new Jupyter notebook or Python script named
# prompt_engineering_lab.ipynb
# Set up your OpenAI client and create helper functions for testing (running multiple calls to OpenAI at the same time)
# Expected outcome: You have a working environment with helper functions ready to use.

# Checkpoint: Run the setup code and verify you get the success message.


import os
from openai import OpenAI
import json
from collections import Counter
import time
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Initialize OpenAI client
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

#set default model
MODEL = "gpt-4o-mini"

print("Environment setup complete. Ready to create prompts!")

Environment setup complete. Ready to create prompts!


In [2]:
def run_openai_request(prompt, model=MODEL, temperature=0.1, max_tokens=600):
    """
    Function to run a request to the OpenAI API with the given prompt and model.
    Returns the response text.
    """
    try:
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            temperature=temperature,
            max_tokens=max_tokens,
        )
        return response.choices[0].message.content.strip()
    except Exception as e:
        print(f"Error during OpenAI request: {e}")
        return None

In [3]:
# #Test a single response to ensure everything is working
# response = run_openai_request("Say Hello")
# response

In [4]:
def run_multiple_requests(prompt,n,temp):
    responses_list = []
    for i in range(n):
        response = run_openai_request(prompt, temperature=temp)
        responses_list.append(response)
        # print(f"Response {i+1}: {response}")
        time.sleep(1)  # Sleep for 1 second between requests to avoid rate limits
    return responses_list


In [5]:
# run_multiple_requests("Say Hello", 5,1)

In [6]:
# Step 2: Create Initial Prompts
# What to do:

# Write three basic zero-shot prompts (one for each task type: Sentiment Analysis, 
# Product Description, Data Extraction)
# Test each prompt once to see initial results
# Task 1: Sentiment Analysis

# Create a simple prompt to classify customer messages (assuming call_openai is your helper function):


# Initial simple prompt for sentiment analysis
sentiment_prompt_v1 = """
Classify this customer message: "I love this product! It's exactly what I needed."
"""
 
# Test it once
result = run_openai_request(sentiment_prompt_v1, temperature=0.1)
print("Sentiment Analysis Result:")
print(result)

Sentiment Analysis Result:
This customer message can be classified as **Positive Feedback** or **Customer Satisfaction**.


In [7]:
# Task 2: Product Description Generation

# Create a simple prompt to generate product descriptions:

# Initial simple prompt for product description
product_prompt_v1 = """
Create a product description for a wireless mouse that costs $29.99.
"""
 
# Test it once
result = run_openai_request(product_prompt_v1, temperature=0.1)
print("Product Description Result:")
print(result)

Product Description Result:
**Product Description: Wireless Precision Mouse**

Elevate your computing experience with our Wireless Precision Mouse, designed for seamless navigation and ultimate comfort. Priced at just $29.99, this sleek and stylish mouse combines cutting-edge technology with ergonomic design, making it the perfect companion for both work and play.

**Key Features:**

- **Wireless Freedom:** Say goodbye to tangled cords! Our advanced 2.4GHz wireless technology ensures a stable connection with a range of up to 33 feet, allowing you to move freely without restrictions.

- **Ergonomic Design:** Crafted with your comfort in mind, the Wireless Precision Mouse features a contoured shape that fits perfectly in your hand, reducing strain during long hours of use. The textured grip provides added control, making it ideal for extended gaming sessions or professional tasks.

- **High Precision Tracking:** Equipped with a high-precision optical sensor, this mouse delivers smooth an

In [8]:
# Task 3: Data Extraction

# Create a simple prompt to extract structured data:

# Initial simple prompt for data extraction
extraction_prompt_v1 = """
Extract information from this customer feedback: "I ordered item #12345 on March 15th. The delivery was fast but the packaging was damaged."
"""
 
# Test it once
result = run_openai_request(extraction_prompt_v1, temperature=0.1)
print("Data Extraction Result:")
print(result)

Data Extraction Result:
Here is the extracted information from the customer feedback:

- **Order Number:** #12345
- **Order Date:** March 15th
- **Delivery Speed:** Fast
- **Packaging Condition:** Damaged


In [9]:
# Part 2: Diagnosing Failures - Systematic Testing
# Objective: Run each prompt multiple times (5, 10, 15 iterations) to identify failure 
# patterns and inconsistencies.

# Step 3: Run Prompts 5 Times
# What to do:

# Run each prompt 5 times using a helper function
# Record all responses
# Analyze consistency between prompts
# Expected outcome: You see variations in outputs across the 5 runs. Document what you observe:

# Are responses consistent in format?
# Are responses consistent in content?
# What types of variations do you see?
# Checkpoint: You've run all three prompts 5 times and recorded the results.

In [10]:
sentiment_prompt_v1 = """
Classify this customer message: "I love this product! It's exactly what I needed."
"""

run_multiple_requests(sentiment_prompt_v1, 5, 1)

['The customer message can be classified as positive feedback or a customer compliment.',
 'This customer message can be classified as **positive feedback** or **customer satisfaction**.',
 'Positive feedback.',
 'This customer message can be classified as **Positive Feedback** or **Customer Satisfaction**.',
 'This customer message can be classified as positive feedback.']

In [11]:
product_prompt_v1 = """
Create a product description for a wireless mouse that costs $29.99.
"""

run_multiple_requests(product_prompt_v1, 5, 1)

['**Product Title: Precision Wireless Mouse**\n\n**Price: $29.99**\n\n**Product Description:**\n\nEnhance your productivity and elevate your workspace with our Precision Wireless Mouse, designed for both comfort and efficiency. Say goodbye to tangled cords and experience the freedom of wireless navigation with up to 33 feet of range, allowing you to work, play, or browse from anywhere in the room.\n\n**Key Features:**\n\n- **Ergonomic Design:** Our mouse features a sleek, ergonomic shape that fits comfortably in your hand, reducing strain during long work hours and enhancing overall user experience.\n  \n- **Plug & Play Convenience:** Setup is a breeze! Simply plug the USB receiver into your computer and enjoy a reliable connection within seconds—no software installation required.\n\n- **High Precision Tracking:** Equipped with advanced optical tracking technology, achieve accurate cursor control on virtually any surface. Whether you’re editing photos or scrolling through spreadsheets,

In [12]:
extraction_prompt_v1 = """
Extract information from this customer feedback: "I ordered item #12345 on March 15th. The delivery was fast but the packaging was damaged."
"""
 
run_multiple_requests(extraction_prompt_v1, 5, 1)

['Here is the extracted information from the customer feedback:\n\n- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Packaging Condition**: Damaged',
 'Here’s the extracted information from the customer feedback:\n\n- Order Number: #12345\n- Order Date: March 15th\n- Delivery Speed: Fast\n- Packaging Condition: Damaged',
 '- **Item Ordered**: #12345\n- **Order Date**: March 15th\n- **Delivery**: Fast\n- **Packaging Condition**: Damaged',
 'Here’s the extracted information from the customer feedback:\n\n- **Order Number:** #12345\n- **Order Date:** March 15th\n- **Delivery Speed:** Fast\n- **Packaging Condition:** Damaged',
 '- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Issue**: Damaged packaging']

In [13]:
# Step 4: Run Prompts 10 Times
# What to do:

# Run each prompt 10 times
# Compare results with the 5-run test
# Identify if more runs reveal additional failure patterns
# Expected outcome: You notice that with more runs, consistency may decrease or new failure patterns emerge.

# Checkpoint: You've completed 10-run tests for all prompts.



In [14]:
run_multiple_requests(sentiment_prompt_v1, 10, 1)


['This customer message can be classified as **Positive Feedback / Satisfaction**.',
 'The customer message can be classified as **Positive Feedback**.',
 'This customer message can be classified as **positive feedback** or **satisfaction**.',
 'This customer message can be classified as "Positive Feedback" or "Satisfaction."',
 'The customer message can be classified as "Positive Feedback" or "Customer Satisfaction."',
 'The customer message can be classified as "Positive Feedback" or "Customer Satisfaction."',
 'The customer message can be classified as **Positive Feedback**.',
 'This customer message can be classified as positive feedback or a positive review.',
 'This customer message can be classified as positive feedback.',
 'The customer message can be classified as "Positive Feedback" or "Satisfaction."']

In [15]:
run_multiple_requests(product_prompt_v1, 10, 1)


["**Product Name:** SwiftClick Wireless Mouse\n\n**Price:** $29.99\n\n**Product Description:**\n\nElevate your computing experience with the SwiftClick Wireless Mouse, designed for functionality, precision, and style. Whether you're a passionate gamer, a busy professional, or someone who simply enjoys browsing the web, this sleek and ergonomic mouse ensures comfort while enhancing your productivity.\n\n**Key Features:**\n\n- **Wireless Freedom:** Enjoy the convenience of a clutter-free workspace with the latest 2.4GHz wireless technology. Experience smooth, lag-free navigation at a range of up to 33 feet, making multitasking easier than ever.\n\n- **Ergonomic Design:** Crafted to fit comfortably in your hand, the SwiftClick minimizes strain and fatigue, allowing for extended use without discomfort. The textured surface provides a secure grip for both left and right-handed users.\n\n- **Precision Performance:** Equipped with a high-precision optical sensor, this mouse offers customizabl

In [16]:
run_multiple_requests(extraction_prompt_v1, 10, 1)

['Here’s the extracted information from the customer feedback:\n\n- **Order Item**: #12345\n- **Order Date**: March 15th\n- **Delivery**: Fast\n- **Packaging Condition**: Damaged',
 '- **Order Number:** #12345\n- **Order Date:** March 15th\n- **Delivery Speed:** Fast\n- **Issue:** Damaged packaging',
 'Here is the extracted information from the customer feedback:\n\n- **Order Item Number:** #12345\n- **Order Date:** March 15th\n- **Delivery Speed:** Fast\n- **Packaging Condition:** Damaged',
 'Here is the extracted information from the customer feedback:\n\n- **Item Ordered**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Packaging Condition**: Damaged',
 'Here is the extracted information from the customer feedback:\n\n- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Packaging Condition**: Damaged',
 'Here is the extracted information from the customer feedback:\n\n- Order number: #12345\n- Order date: March 15th\n- Deli

In [17]:
# Step 5: Run Prompts 15 Times and Create Failure Analysis
# What to do:

# Run each prompt 15 times
# Create a comprehensive failure analysis table
# Document
# specific
# failure patterns for each prompt
# Expected outcome: You have a clear picture of how each prompt fails:

# Low consistency percentages
# Multiple unique responses
# Specific failure patterns documented
# Checkpoint: You've completed 15-run tests and created a failure analysis document.



In [18]:
run_multiple_requests(sentiment_prompt_v1, 15, 1)

['This customer message can be classified as **Positive Feedback**.',
 'The customer message can be classified as positive feedback or a customer compliment.',
 'This customer message can be classified as positive feedback or a product review.',
 'The customer message can be classified as **Positive Feedback** or **Positive Review**.',
 'The customer message can be classified as "Positive Feedback" or "Customer Satisfaction."',
 'This customer message can be classified as **positive feedback** or **customer satisfaction**.',
 'This customer message can be classified as **positive feedback** or a **satisfied customer review**.',
 'Positive Feedback',
 'The customer message can be classified as **positive feedback** or **satisfaction**.',
 'This customer message can be classified as **positive feedback** or **customer satisfaction**.',
 'The customer message can be classified as "Positive Feedback" or "Satisfied Customer."',
 'This customer message can be classified as **positive feedbac

In [19]:
run_multiple_requests(product_prompt_v1, 15, 1)

["**Product Name: SleekConnect Wireless Mouse**\n\n**Price: $29.99**\n\n**Product Description:**\n\nEnhance your workflow with the SleekConnect Wireless Mouse, the perfect blend of style, comfort, and functionality at an unbeatable price. Designed for both casual users and professionals, this sleek and modern mouse delivers a seamless experience, allowing you to navigate effortlessly through your tasks.\n\n**Key Features:**\n\n- **Cutting-Edge Wireless Technology:** Say goodbye to tangled wires and enjoy a reliable, fast connection with advanced 2.4GHz wireless technology, ensuring smooth performance up to 33 feet away from your device.\n\n- **Ergonomic Design:** Designed to fit comfortably in your hand, the SleekConnect Wireless Mouse reduces wrist strain and provides long-lasting comfort, making it ideal for extended use, whether you're working or gaming.\n\n- **Adjustable Sensitivity:** Customize your tracking speed with adjustable DPI settings (800/1200/1600), allowing you to switc

In [20]:
run_multiple_requests(extraction_prompt_v1, 15, 1)

['- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Issue**: Damaged packaging',
 'Here is the extracted information from the customer feedback:\n\n- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Packaging Condition**: Damaged',
 'Here is the extracted information from the customer feedback:\n\n- **Order Number:** 12345\n- **Order Date:** March 15th\n- **Delivery Speed:** Fast\n- **Packaging Condition:** Damaged',
 '- **Item Ordered**: #12345\n- **Order Date**: March 15th\n- **Delivery Time**: Fast\n- **Packaging Condition**: Damaged',
 '- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Packaging Condition**: Damaged',
 'Customer Feedback Summary:\n\n- Order Number: #12345\n- Order Date: March 15th\n- Delivery: Fast\n- Packaging Condition: Damaged',
 '- **Order Number**: #12345\n- **Order Date**: March 15th\n- **Delivery Speed**: Fast\n- **Issue**: Damaged packaging

In [21]:
# Part 3: Iteration 1 - Rewriting Simple Prompts
# Objective: Improve prompts based on failure analysis by adding clarity, format requirements, and constraints.

# Step 6: Improve Sentiment Analysis Prompt
# What to do:

# Rewrite the sentiment analysis prompt with:
# Clearer instructions
# Explicit output format requirement
# Constraints (single word response)
# Expected outcome: Improved consistency compared to version 1.

sentiment_prompt_v2 = """
Classify this customer message in one word either Positive or Negative : "I love this product! It's exactly what I needed."
"""

run_multiple_requests(sentiment_prompt_v2, 15, 1) 


['Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive',
 'Positive']

In [22]:
# Step 7: Improve Product Description Prompt
# What to do:

# Rewrite the product description prompt with:
# Clear structure requirements
# Length constraints
# Style guidelines
# Expected outcome: More consistent length and style, though content may still vary.


product_prompt_v2 = """
Create a product description for a wireless mouse that costs $29.99.
The description should be between 50 and 100 words and use a professional tone.
Focus only on the product's features (ergonomic design, long battery life, great price, reliable connectivity of 2.4Ghz) and benefits, avoiding any personal opinions or unrelated information.
"""

run_multiple_requests(product_prompt_v2, 10, 1)


['Introducing our Wireless Mouse, priced at just $29.99, designed for effortless productivity and comfort. Featuring an ergonomic design, this mouse ensures optimal hand positioning, reducing strain during extended use. Enjoy long-lasting performance with a battery life that keeps you working without interruption. The reliable 2.4GHz connectivity guarantees a stable and fast response, allowing for seamless navigation. Ideal for both professional and casual users, this wireless mouse combines functionality and value in a sleek package. Upgrade your workspace today with this essential accessory that enhances efficiency and comfort.',
 'Introducing our Wireless Mouse, designed for optimal comfort and efficiency. Featuring an ergonomic design, it reduces strain during extended use, ensuring a pleasant experience for users of all styles. Enjoy long-lasting functionality with an impressive battery life, allowing for uninterrupted productivity. With reliable 2.4GHz connectivity, you can expec

In [23]:
# Step 8: Improve Data Extraction Prompt
# What to do:

# Rewrite the data extraction prompt with:
# Structured output format
# Specific fields to extract
# Clear format requirements
# Expected outcome: More structured output, though JSON parsing may still have issues.

# Checkpoint: You've improved all three prompts and tested them. Compare consistency with version 1.


extraction_prompt_v2 = """
Extract the following information from this customer feedback: "I ordered item #12345 on March 15th. The delivery was fast but the packaging was damaged."
- Item ID:followed by 5 digits in the format of #xxxxx
- Order Date:followed by the date in the format of DD/MM/YYYY
- Delivery Status: followed by either word Delivered/Not Delivered
- Packaging Condition:followed by the word Good/Damaged
"""
 
run_multiple_requests(extraction_prompt_v2, 5, 1)


['- Item ID: #12345\n- Order Date: 15/03/2023\n- Delivery Status: Delivered\n- Packaging Condition: Damaged',
 '- Item ID: #12345\n- Order Date: 15/03/2023\n- Delivery Status: Delivered\n- Packaging Condition: Damaged',
 '- Item ID: #12345\n- Order Date: 15/03/2023\n- Delivery Status: Delivered\n- Packaging Condition: Damaged',
 '- Item ID: #12345\n- Order Date: 15/03/2023\n- Delivery Status: Delivered\n- Packaging Condition: Damaged',
 '- Item ID: #12345\n- Order Date: 15/03/2023\n- Delivery Status: Delivered\n- Packaging Condition: Damaged']

In [24]:
# Part 4: Iteration 2 - Adding Structure & Constraints
# Objective: Add few-shot examples, Chain-of-Thought reasoning, and advanced structure to dramatically improve consistency.

# Step 9: Add Few-Shot Examples to Sentiment Analysis
# What to do:

# Add 2-3 examples showing the exact format you want
# Test the few-shot prompt 15 times
# Expected outcome: Significantly improved consistency (likely 80%+).

sentiment_prompt_v3 = """
Classify this customer message as either Positive or Negative: "I love this product! It's exactly what I needed."
Format: the message in quotes, followed by " - " and exactly one word: Positive or Negative
Example: "I love this product!" - Positive
Example: "This product is terrible." - Negative
"""

run_multiple_requests(sentiment_prompt_v3, 15, 1)

['"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positive',
 '"I love this product! It\'s exactly what I needed." - Positi

In [25]:
# Step 10: Add Chain-of-Thought to Data Extraction
# What to do:

# Add Chain-of-Thought reasoning for complex extraction tasks
# Test the CoT prompt 15 times
# Expected outcome: Improved reasoning quality and more consistent extraction.

extraction_prompt_v3 = """
Extract the following information from this customer feedback: "I ordered item #12345 on March 15th. The delivery was fast but the packaging was damaged."
First identify each field in the text step by step, 
then output the final structured result in the following order
1.Item ID:followed by 5 digits in the format of #xxxxx
2.Order Date:followed by the date in the format of DD/MM/YYYY
3.Delivery Status: followed by either word Delivered/Not Delivered
4.Packaging Condition:followed by the word Good/Damaged
"""
run_multiple_requests(extraction_prompt_v3, 15,1) 


['Let\'s break down the customer feedback step by step to identify each field.\n\n1. **Item ID**: The feedback mentions "item #12345". Here, we identify the item ID as "12345".\n2. **Order Date**: The order date is provided as "March 15th". We convert this to a numeric format: March 15, 2023 will be represented as "15/03/2023".\n3. **Delivery Status**: The text states "The delivery was fast". Although it doesn\'t explicitly state "delivered", we can infer that the delivery was completed on time, thus we classify the delivery status as "Delivered".\n4. **Packaging Condition**: The feedback mentions "the packaging was damaged", which clearly indicates that the packaging condition is "Damaged".\n\nNow, let\'s compile the structured result based on the identified fields:\n\n1. Item ID: #12345\n2. Order Date: 15/03/2023\n3. Delivery Status: Delivered\n4. Packaging Condition: Damaged\n\nFinally, here\'s the complete structured output:\n\n```\nItem ID: #12345\nOrder Date: 15/03/2023\nDelivery

In [26]:
# Step 11: Add Few-Shot Examples and Structure to Product Description
# What to do:

# Combine few-shot examples with structured output format
# Test the combined approach 15 times
# Expected outcome: High consistency with proper format and style.

# Checkpoint: You've implemented few-shot and CoT techniques. Compare all version 3 results with version 1.


product_prompt_v3 = """
Create a product description for a wireless mouse that costs $29.99.
The description should be around 50 words and use a professional tone.
Focus only on the product's features and describe them in the following order (ergonomic design, long battery life, reliable connectivity of 2.4GHz, great value for money), avoiding any personal opinions or unrelated information.
For example:
    Introducing our Wireless Mouse, designed with an ergonomic shape for comfortable use during long hours.
Enjoy extended productivity with a long battery life, ensuring uninterrupted sessions.
The 2.4GHz reliable connectivity offers a seamless experience without lag.
Priced at just $29.99, this mouse delivers exceptional value for money,
combining quality and performance.
"""

run_multiple_requests(product_prompt_v3, 15, 1)

['Introducing our Wireless Mouse, crafted with an ergonomic design to enhance comfort during extended use. Experience uninterrupted productivity with its long battery life. The 2.4GHz reliable connectivity ensures a seamless, lag-free experience. Priced at $29.99, this mouse offers exceptional value for money, combining superior quality with impressive performance.',
 'Introducing our Wireless Mouse, expertly crafted with an ergonomic design for optimal comfort during extended use. Experience extended productivity with impressive long battery life, minimizing the need for frequent replacements. The 2.4GHz reliable connectivity ensures a smooth, lag-free experience. Priced at $29.99, this mouse offers outstanding value for money, merging quality and functionality.',
 'Introducing our Wireless Mouse, expertly crafted with an ergonomic design for enhanced comfort during prolonged use. Benefit from an impressive long battery life, allowing for extended productivity without interruptions. T

## Failure Analysis & Consistency Report

To quantify the improvement from v1 → v2 → v3, every prompt version is run **15 times** and scored on two metrics:

- **Exact-match consistency** — the share of runs that produced the single most common response. Strict, and the right metric for classification (sentiment) and extraction, where a production system needs byte-identical, parseable output.
- **Format adherence** — the share of runs that satisfy the task's format rules (single-line label, ~50-word plain prose, all required field names). This is the fair metric for the generative product-description task, where wording is *expected* to vary but structure must not. The format rules are the **final production spec**, applied uniformly to all versions, so earlier versions are scored against the same target the final prompt must hit.

Every version is tested at the **same temperature (1.0)**, so the only variable changing between versions is the prompt itself — any consistency gain is attributable purely to prompt design, not to temperature tuning.

In [27]:
# Failure analysis helpers
# Two metrics per prompt:
#  - exact-match consistency: % of runs equal to the most common response (strict; suits classification/extraction)
#  - format adherence: % of runs that satisfy the task's format rules (suits generative tasks where wording may vary)

import re

def word_count(text):
    return len(text.split())

def sentiment_format_ok(r):
    # Expected: one line ending with a single allowed label word
    return r is not None and len(r.strip().splitlines()) == 1 and \
        bool(re.search(r'\b(Positive|Negative|Neutral)\.?$', r.strip()))

def product_format_ok(r):
    # Expected: ~50 words (40-70 tolerance), plain prose without markdown headers/bullets
    return r is not None and 40 <= word_count(r) <= 70 and "**" not in r and "\n-" not in r

def extraction_format_ok(r):
    # Expected: all four required field names present, exactly as specified
    required = ["Item ID", "Order Date", "Delivery Status", "Packaging Condition"]
    return r is not None and all(field in r for field in required)

def analyze_consistency(responses, format_checker):
    counts = Counter(responses)
    most_common_resp, most_common_n = counts.most_common(1)[0]
    n = len(responses)
    return {
        "runs": n,
        "unique_responses": len(counts),
        "exact_match_consistency_pct": round(most_common_n / n * 100, 1),
        "format_adherence_pct": round(sum(bool(format_checker(r)) for r in responses) / n * 100, 1),
        "most_common": most_common_resp,
    }

print("Analysis helpers ready.")

Analysis helpers ready.


In [28]:
# Collect 15-run results for every task x version, all at the same temperature (1) for a fair comparison.
# This gives one comparable dataset for the failure analysis table.

test_matrix = {
    ("Sentiment", "v1"): (sentiment_prompt_v1, 1),
    ("Sentiment", "v2"): (sentiment_prompt_v2, 1),
    ("Sentiment", "v3"): (sentiment_prompt_v3, 1),
    ("Product", "v1"): (product_prompt_v1, 1),
    ("Product", "v2"): (product_prompt_v2, 1),
    ("Product", "v3"): (product_prompt_v3, 1),
    ("Extraction", "v1"): (extraction_prompt_v1, 1),
    ("Extraction", "v2"): (extraction_prompt_v2, 1),
    ("Extraction", "v3"): (extraction_prompt_v3, 1),
}

all_results = {}
for (task, version), (prompt, temp) in test_matrix.items():
    print(f"Running {task} {version} - 15 runs at temperature {temp}...")
    all_results[(task, version)] = run_multiple_requests(prompt, 15, temp)

print("Done collecting results.")

Running Sentiment v1 - 15 runs at temperature 1...


Running Sentiment v2 - 15 runs at temperature 1...


Running Sentiment v3 - 15 runs at temperature 1...


Running Product v1 - 15 runs at temperature 1...


Running Product v2 - 15 runs at temperature 1...


Running Product v3 - 15 runs at temperature 1...


Running Extraction v1 - 15 runs at temperature 1...


Running Extraction v2 - 15 runs at temperature 1...


Running Extraction v3 - 15 runs at temperature 1...


Done collecting results.


In [29]:
# Failure Analysis Report: consistency table across all tasks and versions

format_checkers = {
    "Sentiment": sentiment_format_ok,
    "Product": product_format_ok,
    "Extraction": extraction_format_ok,
}

report = {}
print("| Task       | Version | Exact-match consistency | Format adherence | Unique responses (of 15) |")
print("|------------|---------|-------------------------|------------------|--------------------------|")
for (task, version), responses in all_results.items():
    stats = analyze_consistency(responses, format_checkers[task])
    report[(task, version)] = stats
    print(f"| {task:<10} | {version:<7} | {stats['exact_match_consistency_pct']:>22}% | {stats['format_adherence_pct']:>15}% | {stats['unique_responses']:>24} |")

print("\n--- v1 vs v3 improvement ---")
for task in ["Sentiment", "Product", "Extraction"]:
    v1, v3 = report[(task, "v1")], report[(task, "v3")]
    print(f"{task}: exact-match {v1['exact_match_consistency_pct']}% -> {v3['exact_match_consistency_pct']}% | "
          f"format adherence {v1['format_adherence_pct']}% -> {v3['format_adherence_pct']}% | "
          f"unique responses {v1['unique_responses']} -> {v3['unique_responses']}")

| Task       | Version | Exact-match consistency | Format adherence | Unique responses (of 15) |
|------------|---------|-------------------------|------------------|--------------------------|
| Sentiment  | v1      |                   13.3% |             0.0% |                       11 |
| Sentiment  | v2      |                  100.0% |           100.0% |                        1 |
| Sentiment  | v3      |                  100.0% |           100.0% |                        1 |
| Product    | v1      |                    6.7% |             0.0% |                       15 |
| Product    | v2      |                    6.7% |             0.0% |                       15 |
| Product    | v3      |                    6.7% |           100.0% |                       15 |
| Extraction | v1      |                   20.0% |             0.0% |                       12 |
| Extraction | v2      |                   93.3% |           100.0% |                        2 |
| Extraction | v3      |      

### Documented Failure Patterns (from the 5/10/15-run tests, all at temperature 1)

**Task 1 — Sentiment Analysis** (exact-match: v1 13.3% → v2 100% → v3 100%)
- **v1 failures:** Never returned a single clean label (11 unique responses in 15 runs). Responses were full sentences ("This customer message can be classified as ..."), often proposed *two* invented categories at once ("Positive Feedback" or "Customer Satisfaction"), and the markdown emphasis, quoting style, and wording changed run to run. A downstream check like `if sentiment == "positive"` would fail on every run.
- **v2 fix:** Constraining the output to a single word from an allowed set was enough for 100% exact-match consistency — even at temperature 1.
- **v3 fix:** Few-shot examples additionally lock the output *shape* (message + " - " + label), keeping 100% consistency while making the format explicit rather than implied.

**Task 2 — Product Description** (format adherence: v1 0% → v2 0% → v3 100%)
- **v1 failures:** Wildly variable length (200–400+ words), invented product names (SwiftConnect, Aero Glide) and invented specs (DPI settings, 33-ft range, AA batteries) that differ per run, and heavy markdown structure with headers/bullets that changed every run.
- **v2 fix:** A 50–100-word limit, tone requirement, and fixed feature list stopped the hallucinated specs — but v2 scores 0% here because the format checker enforces the *final production spec* (~50 words, 40–70 tolerance) uniformly across versions, and v2's looser limit produced ~90-word outputs.
- **v3 fix:** A worked example (few-shot) + a required feature *order* pinned length and structure to the target spec: 100% format adherence. Exact-match stays at 6.7% by nature — a generative task at temperature 1 is *supposed* to vary in wording — which is why format adherence is the right metric for this task.

**Task 3 — Data Extraction** (exact-match: v1 20% → v2 93.3% → v3 6.7%; format adherence: 0% → 100% → 100%)
- **v1 failures:** The model invented its own schema each run (12 unique responses) — field names varied ("Order Number" / "Item Number" / "Order ID"; "Delivery Speed" / "Delivery Experience" / "Issue"), preamble sentences appeared in some runs and not others, and values were unnormalized ("March 15th", "Damaged packaging" vs "Damaged").
- **v2 fix:** An explicit field list with exact names and value formats (DD/MM/YYYY, controlled vocabularies) → 93.3% exact-match, the best strict consistency of any prompt at temperature 1.
- **v3 trade-off (an honest finding):** Adding Chain-of-Thought kept format adherence at 100% — the final structured block is always present and parseable — but *destroyed* exact-match consistency (6.7%), because the model prints its step-by-step reasoning before the result and that reasoning text varies freely between runs. Lesson: CoT improves reasoning reliability but adds variable text; in production you would ask the model to reason *silently* or parse only the final block. Both v2 and v3 also share a residual bug: they confidently output the year "2023" even though the source text contains no year (fixed in Part 5 with a "Not Mentioned" rule).


In [30]:
# Part 5: Tuning for Different Tasks
# Objective: Adapt your prompts for variations of each task type and document what works best.

# Step 12: Create Task Variations
# What to do:

# Create variations of each task
# Test your best prompts on these variations
# Document what works and what needs adjustment
# Expected outcome: You understand how your prompts perform on different variations and can identify what needs tuning.



In [31]:
# Variation 1: Sentiment analysis on different message types (template based on sentiment_prompt_v3)
# The template turns the v3 prompt into a reusable function and adds a Neutral label,
# since real customer messages are not always clearly positive or negative.

def make_sentiment_prompt(message):
    return f"""
Classify this customer message as Positive, Negative, or Neutral: "{message}"
Format: the message in quotes, followed by " - " and exactly one word: Positive, Negative, or Neutral
Example: "I love this product!" - Positive
Example: "This product is terrible." - Negative
Example: "The mouse arrived on Tuesday." - Neutral
"""

sentiment_variations = [
    "The delivery took two weeks and the box was crushed. Very disappointed.",  # clearly negative
    "The mouse works as described.",                                            # neutral / factual
    "Great mouse, but the scroll wheel already squeaks.",                       # mixed sentiment
]

for msg in sentiment_variations:
    print("MESSAGE:", msg)
    for r in run_multiple_requests(make_sentiment_prompt(msg), 5, 1):
        print("   ", r)
    print()

MESSAGE: The delivery took two weeks and the box was crushed. Very disappointed.


    "The delivery took two weeks and the box was crushed. Very disappointed." - Negative
    "The delivery took two weeks and the box was crushed. Very disappointed." - Negative
    "The delivery took two weeks and the box was crushed. Very disappointed." - Negative
    "The delivery took two weeks and the box was crushed. Very disappointed." - Negative
    "The delivery took two weeks and the box was crushed. Very disappointed." - Negative

MESSAGE: The mouse works as described.


    "The mouse works as described." - Positive
    "The mouse works as described." - Positive
    "The mouse works as described." - Positive
    "The mouse works as described." - Positive
    "The mouse works as described." - Positive

MESSAGE: Great mouse, but the scroll wheel already squeaks.


    "Great mouse, but the scroll wheel already squeaks." - Negative
    "Great mouse, but the scroll wheel already squeaks." - Negative
    "Great mouse, but the scroll wheel already squeaks." - Neutral
    "Great mouse, but the scroll wheel already squeaks." - Negative
    "Great mouse, but the scroll wheel already squeaks." - Negative



In [32]:
# Variation 2: Product descriptions for different products (template based on product_prompt_v3)

def make_product_prompt(product, price, features):
    return f"""
Create a product description for {product} that costs {price}.
The description should be around 50 words and use a professional tone.
Focus only on the product's features and describe them in the following order ({features}), avoiding any personal opinions or unrelated information.
For example:
    Introducing our Wireless Mouse, designed with an ergonomic shape for comfortable use during long hours.
Enjoy extended productivity with a long battery life, ensuring uninterrupted sessions.
The 2.4GHz reliable connectivity offers a seamless experience without lag.
Priced at just $29.99, this mouse delivers exceptional value for money,
combining quality and performance.
"""

product_variations = [
    ("bluetooth headphones", "$49.99", "active noise cancellation, 30-hour battery life, Bluetooth 5.3 connectivity, great value for money"),
    ("a mechanical keyboard", "$89.99", "tactile switches, RGB backlighting, detachable USB-C cable, great value for money"),
]

for product, price, features in product_variations:
    print("PRODUCT:", product, price)
    for r in run_multiple_requests(make_product_prompt(product, price, features), 5, 1):
        print(r)
        print("   ---")
    print()

PRODUCT: bluetooth headphones $49.99


Introducing our Bluetooth Headphones, featuring advanced active noise cancellation for an immersive listening experience. Enjoy up to 30 hours of battery life for extended use without interruptions. Equipped with Bluetooth 5.3 connectivity, these headphones offer a stable and efficient wireless connection. Priced at $49.99, they provide outstanding value for money.
   ---
Introducing our Bluetooth Headphones, equipped with advanced active noise cancellation for an immersive listening experience. Enjoy up to 30 hours of battery life for uninterrupted listening. Featuring Bluetooth 5.3 connectivity, these headphones ensure a stable and efficient connection. At just $49.99, they provide outstanding value for money without compromising quality.
   ---
Introducing our Bluetooth Headphones, equipped with advanced active noise cancellation for an immersive listening experience. Enjoy up to 30 hours of battery life, providing extended usage without frequent recharging. Featuring Bluetooth 5.3 

Discover our Mechanical Keyboard featuring tactile switches that provide satisfying feedback for enhanced typing accuracy. The customizable RGB backlighting allows for dynamic illumination, adding a personal touch to your setup. A detachable USB-C cable ensures easy transport and connectivity. Priced at $89.99, this keyboard represents great value for money, blending quality and performance seamlessly.
   ---
Introducing our Mechanical Keyboard, featuring tactile switches for precise feedback and enhanced typing experience. Immerse yourself in customizable RGB backlighting, offering a spectrum of vibrant colors for a personalized setup. The convenient detachable USB-C cable ensures easy transport and connectivity. Priced at $89.99, this keyboard delivers outstanding value for money, merging quality with performance.
   ---
Introducing our Mechanical Keyboard, featuring tactile switches for enhanced typing feedback and precision. The vibrant RGB backlighting offers customizable illumina

In [33]:
# Variation 3: Data extraction on different feedback texts (template based on extraction_prompt_v3)
# Includes a text with MISSING fields to test how the prompt copes; "Not Mentioned" is allowed
# so the model does not invent values (the year-guessing problem seen in v2/v3).

def make_extraction_prompt(feedback):
    return f"""
Extract the following information from this customer feedback: "{feedback}"
First identify each field in the text step by step,
then output the final structured result in the following order.
If a field is not mentioned in the text, write "Not Mentioned" instead of guessing.
1. Item ID: followed by 5 digits in the format #xxxxx
2. Order Date: followed by the date in the format DD/MM/YYYY (use "Not Mentioned" if the year or date is absent)
3. Delivery Status: followed by either word Delivered/Not Delivered
4. Packaging Condition: followed by the word Good/Damaged
"""

extraction_variations = [
    "I ordered item #98765 on April 2nd 2024. It still has not arrived and customer support was unhelpful.",
    "Item #54321 arrived in perfect packaging. Great service!",  # no date at all
]

for feedback in extraction_variations:
    print("FEEDBACK:", feedback)
    for r in run_multiple_requests(make_extraction_prompt(feedback), 5, 1):
        print(r)
        print("   ---")
    print()

FEEDBACK: I ordered item #98765 on April 2nd 2024. It still has not arrived and customer support was unhelpful.


Let's break down the feedback step by step:

1. **Item ID:** The text includes "item #98765", which means the Item ID is #98765.
  
2. **Order Date:** The text indicates "ordered on April 2nd 2024". We can convert this to the required format as 02/04/2024.

3. **Delivery Status:** The feedback states "It still has not arrived", indicating that the delivery status is Not Delivered.

4. **Packaging Condition:** The text does not mention anything about the packaging condition, so this will be noted as Not Mentioned.

Now, let's compile these findings into the final structured result.

Final Structured Result:
1. Item ID: #98765
2. Order Date: 02/04/2024
3. Delivery Status: Not Delivered
4. Packaging Condition: Not Mentioned
   ---
Let's extract the required fields step by step.

1. **Item ID**: The text mentions "item #98765." Therefore, the Item ID is #98765.
  
2. **Order Date**: The feedback states "on April 2nd 2024." We need to convert this to the format DD/MM/YYYY. April 2nd is the 

Let's extract the information step by step from the customer feedback: "Item #54321 arrived in perfect packaging. Great service!"

1. **Item ID:**
   - The text mentions "Item #54321", so the Item ID is #54321.

2. **Order Date:**
   - The text does not mention any date, so Order Date is "Not Mentioned".

3. **Delivery Status:**
   - The text states "arrived", which implies that the item has been delivered. Therefore, Delivery Status is Delivered.

4. **Packaging Condition:**
   - The text mentions "in perfect packaging", which indicates that the packaging is in good condition. Thus, Packaging Condition is Good.

Now, let's put this all together in the required structured format.

Final structured result:
1. Item ID: #54321
2. Order Date: Not Mentioned
3. Delivery Status: Delivered
4. Packaging Condition: Good
   ---
Let's break down the customer feedback step by step:

1. **Item ID:** The text includes "Item #54321", suggesting the item ID is #54321. 
   
   **Item ID:** #54321

2. **

### Step 12 findings — what works and what needs adjustment

- **Sentiment template:** generalizes perfectly to clearly negative messages (5/5 "Negative") and holds its format on every variation. Two real weaknesses surfaced: the factual, opinion-free message ("The mouse works as described.") was classified **Positive 5/5 despite the Neutral option** — the model reads "works as described" as satisfaction — and the mixed message ("Great mouse, but the scroll wheel already squeaks.") split 4× Negative / 1× Neutral. Single-label classification hides nuance; a production version would need better Neutral few-shot examples closer to real cases, and possibly a "Mixed" label.
- **Product template:** turning v3 into a function with `product / price / features` parameters worked with no other changes — structure, ~50-word length, and feature order carried over to headphones and a keyboard. The few-shot example does the heavy lifting.
- **Extraction template:** the fixed schema held on new feedback texts, and the added rule to output **"Not Mentioned"** for absent fields worked exactly as intended (5/5 runs on the no-date feedback said "Order Date: Not Mentioned" instead of inventing a year — fixing the hallucinated "2023" seen in v2/v3). The CoT reasoning remains verbose and varies between runs, confirming the Step 10 finding: parse the final block, or ask for silent reasoning.


In [34]:
# Step 13: Final Evaluation and Comparison
# The v3 prompts were already run 15 times each in the Failure Analysis section (all_results / report).
# This cell condenses those numbers into the final comparison report.

print("FINAL COMPARISON REPORT (15 runs per version)\n")
print(f"{'Task':<12}{'Metric':<28}{'v1':>8}{'v2':>8}{'v3':>8}")
print("-" * 64)
for task in ["Sentiment", "Product", "Extraction"]:
    for metric, label in [("exact_match_consistency_pct", "Exact-match consistency %"),
                          ("format_adherence_pct", "Format adherence %"),
                          ("unique_responses", "Unique responses")]:
        row = [report[(task, v)][metric] for v in ["v1", "v2", "v3"]]
        print(f"{task:<12}{label:<28}{row[0]:>8}{row[1]:>8}{row[2]:>8}")
    print("-" * 64)

for task in ["Sentiment", "Product", "Extraction"]:
    v1, v3 = report[(task, "v1")], report[(task, "v3")]
    delta = v3["format_adherence_pct"] - v1["format_adherence_pct"]
    print(f"{task}: format adherence improved by {delta:+.1f} percentage points (v1 -> v3)")

FINAL COMPARISON REPORT (15 runs per version)

Task        Metric                            v1      v2      v3
----------------------------------------------------------------
Sentiment   Exact-match consistency %       13.3   100.0   100.0
Sentiment   Format adherence %               0.0   100.0   100.0
Sentiment   Unique responses                  11       1       1
----------------------------------------------------------------
Product     Exact-match consistency %        6.7     6.7     6.7
Product     Format adherence %               0.0     0.0   100.0
Product     Unique responses                  15      15      15
----------------------------------------------------------------
Extraction  Exact-match consistency %       20.0    93.3     6.7
Extraction  Format adherence %               0.0   100.0   100.0
Extraction  Unique responses                  12       2      15
----------------------------------------------------------------
Sentiment: format adherence improved by +10